File: verification/go/execution_test.go
package agentpayverify

import (
	"context"
	"crypto/ecdsa"
	"crypto/elliptic"
	"crypto/rand"
	"crypto/sha256"
	"encoding/hex"
	"errors"
	"testing"
	"time"

	"github.com/golang-jwt/jwt/v5"
)

func TestExecutionVerifierValidatesBindingsAndConsumesJTIOnce(t *testing.T) {
	t.Parallel()

	privateKey, err := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
	if err != nil {
		t.Fatal(err)
	}
	now := time.Date(2026, time.September, 18, 10, 0, 0, 0, time.UTC)
	resolver := &staticExecutionKeyResolver{key: &privateKey.PublicKey}
	verifier, err := NewExecutionVerifier(ExecutionConfig{
		Issuer: "https://api.agentpay.test", SellerID: "sel_test", RouteID: "rte_test",
		ReplayStore: NewMemoryExecutionReplayStore(), KeyResolver: resolver,
		Clock: func() time.Time { return now },
	})
	if err != nil {
		t.Fatal(err)
	}
	body := []byte(`{"topic":"payments"}`)
	token := signedExecutionToken(t, privateKey, now, body, nil)
	claims, err := verifier.Verify(
		t.Context(), "POST", "/research", body, token, "txn_test",
	)
	if err != nil {
		t.Fatal(err)
	}
	if claims.TransactionID != "txn_test" || claims.JWTID != "xec_test" {
		t.Fatalf("claims = %#v", claims)
	}
	if _, err := verifier.Verify(t.Context(), "POST", "/research", body, token, "txn_test"); !errors.Is(err, ErrReplay) {
		t.Fatalf("replay error = %v", err)
	}
}

func TestExecutionVerifierRejectsEveryMismatchedBinding(t *testing.T) {
	t.Parallel()

	privateKey, err := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
	if err != nil {
		t.Fatal(err)
	}
	now := time.Date(2026, time.September, 18, 10, 0, 0, 0, time.UTC)
	body := []byte(`{"topic":"payments"}`)
	tests := []struct {
		name   string
		method string
		path   string
		body   []byte
		txn    string
		mutate func(jwt.MapClaims)
	}{
		{name: "issuer", method: "POST", path: "/research", body: body, txn: "txn_test", mutate: func(claims jwt.MapClaims) { claims["iss"] = "https://evil.example" }},
		{name: "audience", method: "POST", path: "/research", body: body, txn: "txn_test", mutate: func(claims jwt.MapClaims) { claims["aud"] = "urn:agentpay:seller:sel_other" }},
		{name: "seller", method: "POST", path: "/research", body: body, txn: "txn_test", mutate: func(claims jwt.MapClaims) { claims["sellerId"] = "sel_other" }},
		{name: "route", method: "POST", path: "/research", body: body, txn: "txn_test", mutate: func(claims jwt.MapClaims) { claims["routeId"] = "rte_other" }},
		{name: "method", method: "GET", path: "/research", body: body, txn: "txn_test"},
		{name: "path", method: "POST", path: "/other", body: body, txn: "txn_test"},
		{name: "body", method: "POST", path: "/research", body: []byte(`{"topic":"changed"}`), txn: "txn_test"},
		{name: "transaction header", method: "POST", path: "/research", body: body, txn: "txn_other"},
		{name: "finality", method: "POST", path: "/research", body: body, txn: "txn_test", mutate: func(claims jwt.MapClaims) { claims["paymentFinality"] = "confirmed" }},
		{name: "lifetime", method: "POST", path: "/research", body: body, txn: "txn_test", mutate: func(claims jwt.MapClaims) { claims["exp"] = now.Add(61 * time.Second).Unix() }},
	}
	for _, test := range tests {
		t.Run(test.name, func(t *testing.T) {
			verifier, createErr := NewExecutionVerifier(ExecutionConfig{
				Issuer: "https://api.agentpay.test", SellerID: "sel_test", RouteID: "rte_test",
				ReplayStore: NewMemoryExecutionReplayStore(),
				KeyResolver: &staticExecutionKeyResolver{key: &privateKey.PublicKey},
				Clock:       func() time.Time { return now },
			})
			if createErr != nil {
				t.Fatal(createErr)
			}
			token := signedExecutionToken(t, privateKey, now, body, test.mutate)
			if _, verifyErr := verifier.Verify(t.Context(), test.method, test.path, test.body, token, test.txn); verifyErr == nil {
				t.Fatal("Verify() accepted mismatched execution capability")
			}
		})
	}
}

func signedExecutionToken(
	t *testing.T,
	privateKey *ecdsa.PrivateKey,
	now time.Time,
	body []byte,
	mutate func(jwt.MapClaims),
) string {
	t.Helper()
	digest := sha256.Sum256(body)
	claims := jwt.MapClaims{
		"iss": "https://api.agentpay.test", "aud": "urn:agentpay:seller:sel_test",
		"sub": "txn_test", "sellerId": "sel_test", "routeId": "rte_test",
		"transactionId": "txn_test", "method": "POST", "path": "/research",
		"bodySha256": hex.EncodeToString(digest[:]), "paymentFinality": "finalized",
		"jti": "xec_test", "iat": now.Unix(), "exp": now.Add(45 * time.Second).Unix(),
	}
	if mutate != nil {
		mutate(claims)
	}
	token := jwt.NewWithClaims(jwt.SigningMethodES256, claims)
	token.Header["typ"] = ExecutionCapabilityType
	token.Header["kid"] = "key-1"
	raw, err := token.SignedString(privateKey)
	if err != nil {
		t.Fatal(err)
	}
	return raw
}

type staticExecutionKeyResolver struct {
	key *ecdsa.PublicKey
	err error
}

func (resolver *staticExecutionKeyResolver) ResolveExecutionKey(context.Context, string) (*ecdsa.PublicKey, error) {
	return resolver.key, resolver.err
}

File: packages/merchant-sdk/test/execution.test.mjs
import assert from "node:assert/strict";
import {
  createHash,
  generateKeyPairSync,
  sign as signBytes,
} from "node:crypto";
import test from "node:test";
import {
  createRemoteJwksResolver,
  MemoryFulfillmentStore,
  MemoryExecutionReplayStore,
  MemoryReplayStore,
  MerchantSdkError,
  processAgentPayFulfillment,
  verifyExecutionRequest,
  verifyLegacySandboxRequest,
} from "../dist/index.js";

const now = new Date("2026-09-19T12:00:00Z");

test("verifies a production execution request and blocks capability replay", async () => {
  const { privateKey, publicKey } = generateKeyPairSync("ec", {
    namedCurve: "P-256",
  });
  const body = Buffer.from('{"report":"weekly"}');
  const token = signExecutionCapability({ privateKey, body });
  const options = {
    issuer: "https://api.agentpay.test/",
    sellerId: "sel_test",
    routeId: "rte_test",
    method: "POST",
    path: "/fulfill",
    rawBody: body,
    headers: {
      "X-AgentPay-Execution-Capability": token,
      "X-AgentPay-Transaction-Id": "txn_test",
    },
    keyResolver: async () => publicKey.export({ format: "jwk" }),
    replayStore: new MemoryExecutionReplayStore(),
    now,
  };

  const claims = await verifyExecutionRequest(options);
  assert.equal(claims.transactionId, "txn_test");
  await assert.rejects(verifyExecutionRequest(options), (error) => {
    assert.ok(error instanceof MerchantSdkError);
    assert.equal(error.code, "replay");
    assert.equal(error.status, 409);
    return true;
  });
});

test("rejects execution requests whose raw body binding changed", async () => {
  const { privateKey, publicKey } = generateKeyPairSync("ec", {
    namedCurve: "P-256",
  });
  const signedBody = Buffer.from('{"report":"weekly"}');

  await assert.rejects(
    verifyExecutionRequest({
      issuer: "https://api.agentpay.test",
      sellerId: "sel_test",
      routeId: "rte_test",
      method: "POST",
      path: "/fulfill",
      rawBody: Buffer.from('{"report":"changed"}'),
      headers: {
        "x-agentpay-execution-capability": signExecutionCapability({
          privateKey,
          body: signedBody,
        }),
        "x-agentpay-transaction-id": "txn_test",
      },
      keyResolver: async () => publicKey.export({ format: "jwk" }),
      replayStore: new MemoryExecutionReplayStore(),
      now,
    }),
    (error) =>
      error instanceof MerchantSdkError && error.code === "binding_mismatch",
  );
});

test("maps unavailable execution keys to a safe dependency error", async () => {
  const { privateKey } = generateKeyPairSync("ec", { namedCurve: "P-256" });
  const body = Buffer.from("{}");

  await assert.rejects(
    verifyExecutionRequest({
      issuer: "https://api.agentpay.test",
      sellerId: "sel_test",
      routeId: "rte_test",
      method: "POST",
      path: "/fulfill",
      rawBody: body,
      headers: {
        "x-agentpay-execution-capability": signExecutionCapability({
          privateKey,
          body,
        }),
        "x-agentpay-transaction-id": "txn_test",
      },
      keyResolver: async () => {
        throw new Error("private resolver failure");
      },
      replayStore: new MemoryExecutionReplayStore(),
      now,
    }),
    (error) =>
      error instanceof MerchantSdkError &&
      error.code === "dependency_unavailable" &&
      !error.message.includes("private resolver failure"),
  );
});

test("processes verified JSON through one idempotent merchant adapter call", async () => {
  const { privateKey, publicKey } = generateKeyPairSync("ec", {
    namedCurve: "P-256",
  });
  const body = Buffer.from('{"report":"weekly"}');
  let calls = 0;
  const result = await processAgentPayFulfillment({
    verification: {
      issuer: "https://api.agentpay.test",
      sellerId: "sel_test",
      routeId: "rte_test",
      method: "POST",
      path: "/fulfill",
      rawBody: body,
      headers: {
        "x-agentpay-execution-capability": signExecutionCapability({
          privateKey,
          body,
        }),
        "x-agentpay-transaction-id": "txn_test",
      },
      keyResolver: async () => publicKey.export({ format: "jwk" }),
      replayStore: new MemoryExecutionReplayStore(),
      now,
    },
    fulfillmentStore: new MemoryFulfillmentStore(),
    parseInput(rawBody) {
      return JSON.parse(Buffer.from(rawBody).toString("utf8"));
    },
    adapter: {
      async fulfill(context) {
        calls += 1;
        assert.equal(context.input.report, "weekly");
        return { delivered: true };
      },
    },
  });

  assert.equal(result.disposition, "executed");
  assert.equal(calls, 1);
});

test("exposes the legacy HMAC verifier only through an explicit sandbox name", async () => {
  const secret = "0123456789abcdef0123456789abcdef";
  const body = Buffer.from("sandbox");
  const timestamp = "2026-09-19T12:00:00Z";
  const transactionId = "txn_sandbox";
  const canonical = [
    "agentpay.seller-request.v1",
    timestamp,
    "POST",
    "/fulfill",
    createHash("sha256").update(body).digest("hex"),
    transactionId,
  ].join("\n");
  const { createHmac } = await import("node:crypto");

  await verifyLegacySandboxRequest({
    secret,
    method: "POST",
    path: "/fulfill",
    rawBody: body,
    headers: {
      "x-agentpay-signature": createHmac("sha256", secret)
        .update(canonical)
        .digest("base64"),
      "x-agentpay-timestamp": timestamp,
      "x-agentpay-transaction-id": transactionId,
    },
    replayStore: new MemoryReplayStore(),
    now,
  });
});

test("remote JWKS resolver caches only according to response headers", async () => {
  const { publicKey } = generateKeyPairSync("ec", { namedCurve: "P-256" });
  const jwk = publicKey.export({ format: "jwk" });
  let calls = 0;
  const resolver = createRemoteJwksResolver({
    jwksUrl: "https://api.agentpay.test/.well-known/jwks.json",
    fetch: async () => {
      calls += 1;
      return new Response(
        JSON.stringify({ keys: [{ ...jwk, kid: "key-1", alg: "ES256" }] }),
        {
          status: 200,
          headers: {
            "cache-control": "public, max-age=60",
            "content-type": "application/json",
          },
        },
      );
    },
    now: () => now,
  });

  assert.equal((await resolver("key-1")).kid, "key-1");
  assert.equal((await resolver("key-1")).kid, "key-1");
  assert.equal(calls, 1);
});

function signExecutionCapability({ privateKey, body }) {
  const header = { typ: "agentpay-execution+jwt", alg: "ES256", kid: "key-1" };
  const issuedAt = Math.floor(now.getTime() / 1000);
  const claims = {
    iss: "https://api.agentpay.test",
    aud: "urn:agentpay:seller:sel_test",
    sub: "txn_test",
    sellerId: "sel_test",
    routeId: "rte_test",
    transactionId: "txn_test",
    method: "POST",
    path: "/fulfill",
    bodySha256: createHash("sha256").update(body).digest("hex"),
    paymentFinality: "finalized",
    jti: "xec_merchant_test",
    iat: issuedAt,
    exp: issuedAt + 45,
  };
  const signingInput = `${Buffer.from(JSON.stringify(header)).toString("base64url")}.${Buffer.from(JSON.stringify(claims)).toString("base64url")}`;
  const signature = signBytes("sha256", Buffer.from(signingInput), {
    key: privateKey,
    dsaEncoding: "ieee-p1363",
  });
  return `${signingInput}.${signature.toString("base64url")}`;
}
